presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.h

```diff
@@ -0,0 +1,183 @@
+
+  uint64_t getCompletedRows() override {
+    return completedRows_;
+  }
+
+  std::unordered_map<std::string, velox::RuntimeCounter> runtimeStats()
+      override {
+    return {};
+  }
+
+ private:
+  /// Convert an arrow record batch to Velox RowVector.
+  /// Process only those columns that are present in outputType_
+  velox::RowVectorPtr projectOutputColumns(
+      std::shared_ptr<arrow::RecordBatch> input);
+
+  velox::RowTypePtr outputType_;
+  std::vector<std::string> columnMapping_;
+  std::unique_ptr<arrow::flight::FlightStreamReader> currentReader_;
+  uint64_t completedRows_ = 0;
+  uint64_t completedBytes_ = 0;
+  std::shared_ptr<auth::Authenticator> authenticator_;
+  velox::memory::MemoryPool* pool_;
+  const std::shared_ptr<FlightConfig> flightConfig_;
+};
+
+class ArrowFlightConnector : public velox::connector::Connector {
+ public:
+  explicit ArrowFlightConnector(
+      const std::string& id,
+      std::shared_ptr<const velox::config::ConfigBase> config,
+      const char* authenticatorName = nullptr)
+      : Connector{id},
+        flightConfig_{std::make_shared<FlightConfig>(config)},
+        authenticator_{
+            // auth::getAuthenticatorFactory(config_->authenticatorName())
+            auth::getAuthenticatorFactory(
+                authenticatorName ? authenticatorName
+                                  : flightConfig_->authenticatorName())
+                ->newAuthenticator(config)} {}
+
+  std::unique_ptr<velox::connector::DataSource> createDataSource(
+      const velox::RowTypePtr& outputType,
+      const std::shared_ptr<velox::connector::ConnectorTableHandle>&
+          tableHandle,
+      const std::unordered_map<
+          std::string,
+          std::shared_ptr<velox::connector::ColumnHandle>>& columnHandles,
+      velox::connector::ConnectorQueryCtx* ctx) override {
+    return std::make_unique<FlightDataSource>(
+        outputType,
+        columnHandles,
+        authenticator_,
+        ctx->memoryPool(),
+        flightConfig_);
+  }
+
+  std::unique_ptr<velox::connector::DataSink> createDataSink(
+      velox::RowTypePtr inputType,
+      std::shared_ptr<velox::connector::ConnectorInsertTableHandle>
+          connectorInsertTableHandle,
+      velox::connector::ConnectorQueryCtx* connectorQueryCtx,
+      velox::connector::CommitStrategy commitStrategy) override {
+    VELOX_NYI("Flight connector does not support DataSink");
+  }
+
+ private:
+  const std::shared_ptr<FlightConfig> flightConfig_;
+  const std::shared_ptr<auth::Authenticator> authenticator_;
+};
+
+class ArrowFlightConnectorFactory : public velox::connector::ConnectorFactory {
+ public:
+  static constexpr const char* kArrowFlightConnectorName = "arrow-flight";
+
+  ArrowFlightConnectorFactory() : ConnectorFactory(kArrowFlightConnectorName) {}
+
+  explicit ArrowFlightConnectorFactory(const char* name)
+      : ConnectorFactory{name} {}
+
+  ArrowFlightConnectorFactory(const char* name, const char* authenticatorName)
```

Comment:
> Why do we need this constructor in the first place?

The idea here is to be able to register different connector names which use the same Flight connector code, but use different `Authenticator`s. This would avoid the need to have two separate config properties to fully qualify the constructor; instead of specifying `connector.name` and `arrow-flight.authenticator.name`, you simply specify `connector.name` and are good to go.

This would tie in a little better with the Java version, where authentication needs to be handled by subclasses of `presto-base-arrow-flight`, which are each a unique connector.

Also note that the config `arrow-flight.authenticator.name` is not present in the base Java Flight connector. A concrete connector extension would need to introduce this config explicitly to be able to switch between between C++ `Authenticator`s within the same connector.
https://github.com/prestodb/presto/pull/23032/files#diff-9b8ea7947728279eacfeb69fc0492cb2390c3dd65897bc9148ecf1d021d37332